persistence/src/main/java/de/cyface/persistence/PersistenceLayer.java

```diff
@@ -490,55 +494,126 @@ public void deletePoint3dData(final long measurementIdentifier) {
-            while (cursor.moveToNext()) {
-                final double lat = cursor.getDouble(cursor.getColumnIndex(GeoLocationsTable.COLUMN_LAT));
-                final double lon = cursor.getDouble(cursor.getColumnIndex(GeoLocationsTable.COLUMN_LON));
-                final long timestamp = cursor.getLong(cursor.getColumnIndex(GeoLocationsTable.COLUMN_GEOLOCATION_TIME));
-                final double speed = cursor.getDouble(cursor.getColumnIndex(GeoLocationsTable.COLUMN_SPEED));
-                final float accuracy = cursor.getFloat(cursor.getColumnIndex(GeoLocationsTable.COLUMN_ACCURACY));
-                fullTrack.add(new GeoLocation(lat, lon, timestamp, speed, accuracy));
-            }
+            // Load Events to slice measurements on pause events
+            eventCursor = resolver.query(getEventUri(), null, GeoLocationsTable.COLUMN_MEASUREMENT_FK + "=?",
+                    new String[] {Long.valueOf(measurementIdentifier).toString()},
+                    EventTable.COLUMN_TIMESTAMP + " ASC");
+            Validate.softCatchNullCursor(eventCursor);
 
-            subTracks.add(fullTrack);
-            return subTracks;
+            return loadTracks(geoLocationCursor, eventCursor);
         } finally {
-            if (cursor != null) {
-                cursor.close();
+            if (geoLocationCursor != null) {
+                geoLocationCursor.close();
+            }
+            if (eventCursor != null) {
+                eventCursor.close();
             }
         }
     }
 
+    /**
+     * Loads the {@link Track}s for the provided {@link GeoLocation} cursor sliced using the provided {@link Event}
+     * cursor.
+     *
+     * @param geoLocationCursor The {@code GeoLocation} cursor which points to the locations to be loaded.
+     * @param eventCursor The {@code Event} cursor which points to the events annotation the corresponding
+     *            {@link Measurement}.
+     * @return The {@link Track}s for the corresponding {@code Measurement} loaded.
+     */
+    @NonNull
+    private List<Track> loadTracks(@NonNull final Cursor geoLocationCursor, @NonNull final Cursor eventCursor) {
+        final List<Track> tracks = new ArrayList<>();
+
+        // Slice Tracks before resume events
+        // This helps to allocate GeoLocations which are captured just after pause was hit to the track which just
+        // ended.
+        while (eventCursor.moveToNext()) {
+            final Event.EventType eventType = Event.EventType
+                    .valueOf(eventCursor.getString(eventCursor.getColumnIndex(EventTable.COLUMN_TYPE)));
+            if (eventType != Event.EventType.LIFECYCLE_RESUME) {
+                continue;
+            }
+
+            // get all geolocations captured before this RESUME event
+            final long eventTime = eventCursor.getLong(eventCursor.getColumnIndex(EventTable.COLUMN_TIMESTAMP));
+            final Track track = new Track();
+            while (geoLocationCursor.moveToNext()) {
```

Comment:
I would probably have done this the other way around. Iterate on all the geoLocations and have a second iterator on events, splitting the track at each resume event. I have the feeling that would produce easier to read code. But if it works we may keep this for now.

Reply:
ok